common/src/services/keyConnector.service.ts

```diff
@@ -0,0 +1,86 @@
+import { ApiService } from '../abstractions/api.service';
+import { CryptoService } from '../abstractions/crypto.service';
+import { EnvironmentService } from '../abstractions/environment.service';
+import { KeyConnectorService as KeyConnectorServiceAbstraction } from '../abstractions/keyConnector.service';
+import { LogService } from '../abstractions/log.service';
+import { StorageService } from '../abstractions/storage.service';
+import { UserService } from '../abstractions/user.service';
+
+import { OrganizationUserType } from '../enums/organizationUserType';
+
+import { Utils } from '../misc/utils';
+
+import { SymmetricCryptoKey } from '../models/domain/symmetricCryptoKey';
+
+import { KeyConnectorUserKeyRequest } from '../models/request/keyConnectorUserKeyRequest';
+
+const Keys = {
+    usesKeyConnector: 'usesKeyConnector',
+};
+
+export class KeyConnectorService implements KeyConnectorServiceAbstraction {
+    private usesKeyConnector: boolean = false;
+
+    constructor(private storageService: StorageService, private userService: UserService,
+        private cryptoService: CryptoService, private apiService: ApiService,
+        private environmentService: EnvironmentService, private logService: LogService) { }
+
+    setUsesKeyConnector(usesKeyConnector: boolean) {
+        this.usesKeyConnector = usesKeyConnector;
+        return this.storageService.save(Keys.usesKeyConnector, usesKeyConnector);
+    }
+
+    async getUsesKeyConnector(): Promise<boolean> {
+        return this.usesKeyConnector ??= await this.storageService.get<boolean>(Keys.usesKeyConnector);
+    }
+
+    async userNeedsMigration() {
+        const requiredByOrganization = await this.getManagingOrganization() != null;
+        const userIsNotUsingKeyConnector = !await this.getUsesKeyConnector();
+
+        return requiredByOrganization && userIsNotUsingKeyConnector;
+    }
+
+    async migrateUser() {
+        const organization = await this.getManagingOrganization();
+        const key = await this.cryptoService.getKey();
+
+        try {
+            const keyConnectorRequest = new KeyConnectorUserKeyRequest(key.encKeyB64);
+            await this.apiService.postUserKeyToKeyConnector(organization.keyConnectorUrl, keyConnectorRequest);
+        } catch (e) {
+            throw new Error('Unable to reach key connector');
+        }
+
+        await this.apiService.postConvertToKeyConnector();
+    }
+
+    async getAndSetKey(url?: string) {
+        if (url == null) {
+            url = this.environmentService.getKeyConnectorUrl();
```

Comment:
This environment variable is only used for CLI, I'm not sure whether it's better to use it as the default here, or just require it to be passed in.

Reply:
No-one seems to care so I'm going to leave as is